dipy/align/vector_fields.pyx

```diff
@@ -3065,3 +3065,411 @@ def create_sphere(cnp.npy_intp nslices, cnp.npy_intp nrows,
                 else:
                     s[k,i,j] = 0
     return s
+
+
+cdef void _gradient_3d(floating[:,:,:] img, double[:,:] img_affine_inv,
+                       double[:] img_spacing, double[:,:] domain_affine,
+                       floating[:,:,:,:] out, int[:,:,:] inside):
+    r""" Gradient of a 3D image in physical space coordinates
+
+    Each grid cell (i, j, k) in the sampling domain (determined by
+    out.shape) is mapped to its corresponding physical point (x, y, z) by
+    multiplying domain_affine (its grid-to-space transform) by (i, j, k), then
+    the image is interpolated, at
+
+    P1=(x + h, y, z), Q1=(x - h, y, z)
+    P2=(x, y + h, z), Q2=(x, y - h, z)
+    P3=(x, y, z + h), Q3=(x, y, z - h)
+
+    (by mapping Pi and Qi to the grid using img_affine_inv: the inverse of the
+    grid-to-space transform of img). The displacement parameter h is of
+    magnitude 0.5 (in physical space units), therefore the approximated partial
+    derivatives are given by the difference between the image interpolated at
+    Pi and Qi.
+
+    Parameters
+    ----------
+    img : array, shape (S, R, C)
+        the input volume whose gradient will be computed
+    img_affine_inv : array, shape (4, 4)
+        the space-to-grid transform matrix associated to img
+    img_spacing : array, shape (3,)
+        the spacing between voxels (voxel size along each axis) of the input
+        volume
+    domain_affine : array, shape (4, 4)
```

Comment:
That's correct, every grid has an affine transform associated to it. Now I am using the pattern `x_affine` to refer to the grid-to-space transform associate to grid `x`, and `x_affine_inv` to the space-to-grid transform (sorry, I started using the `domain` naming a long time ago, it's time to stop using it). If you think `x_affine` and `x_affine_inv` are ugly or confusing, or unpractical, I would prefer `x_vox2mm` and `x_mm2vox` (instead of grid2mm and mm2grid). What do you think? Once we agree on a convention I will go ahead and do the necessary changes.